plugins/none-ls/settings.nix

```diff
@@ -0,0 +1,254 @@
+    Defines an `on_exit` callback to run when the null-ls client exits.
+  '';
+
+  root_dir = helpers.defaultNullOpts.mkLuaFn' {
+    pluginDefault = "require('null-ls.utils').root_pattern('.null-ls-root', 'Makefile', '.git')";
+    description = ''
+      Determines the root of the null-ls server. On startup, null-ls will call
+      `root_dir` with the full path to the first file that null-ls attaches to.
+
+      If `root_dir` returns `nil`, the root will resolve to the current working directory.
+    '';
+    example = ''
+      function(fname)
+        return fname:match("my-project") and "my-project-root"
+      end
+    '';
+  };
+
+  root_dir_async = helpers.defaultNullOpts.mkLuaFn' {
+    pluginDefault = null;
+    description = ''
+      Like `root_dir` but also accepts a callback parameter allowing it to be
+      asynchronous. Overwrites `root_dir` when present.
+
+      For a utility that asynchronously finds a matching file, see `utils.root_pattern_async`.
+    '';
+    example = ''
+      function(fname, cb)
+        cb(fname:match("my-project") and "my-project-root")
+      end
+    '';
+  };
+
+  should_attach = helpers.defaultNullOpts.mkLuaFn' {
+    pluginDefault = null;
+    description = ''
+      A user-defined function that controls whether to enable null-ls for a given
+      buffer. Receives `bufnr` as its first argument.
+
+      To cut down potentially expensive calls, null-ls will call `should_attach` after
+      its own internal checks pass, so it's not guaranteed to run on each new buffer.
+    '';
+    example = ''
+      function(bufnr)
+        return not vim.api.nvim_buf_get_name(bufnr):match("^git://")
+      end
+    '';
+  };
+
+  # Not using mkListOf because I want `strLua` instead of `rawLua`
+  sources = helpers.defaultNullOpts.mkNullable' {
+    # TODO: support custom & third-party sources.
+    # Need a "source" submodule type:
+    # https://github.com/nvimtools/none-ls.nvim/blob/main/doc/MAIN.md#sources
+    type = with types; listOf strLua;
+    apply = x: if x == null then null else applyListOfLua x;
+    pluginDefault = null;
+    description = ''
+      The list of sources to enable, should be strings of lua code. Don't use this directly.
+
+      You should use `plugins.none-ls.sources.*.enable` instead.
+    '';
+    visible = false;
+    # TODO: Upstream's description:
+    # description = ''
+    #   Defines a list of sources for null-ls to register.
+    #   Users can add built-in sources (see [BUILTINS]) or custom sources (see [MAIN]).
+    #
+    #   If you've installed an integration that provides its own sources and aren't
+    #   interested in built-in sources, you don't have to define any sources here. The
+    #   integration will register them independently.
+    #
+    #   [BUILTINS]: https://github.com/nvimtools/none-ls.nvim/blob/main/doc/BUILTINS.md
+    #   [MAIN]: https://github.com/nvimtools/none-ls.nvim/blob/main/doc/MAIN.md
+    # '';
```

Comment:
Same, this should be dealt with here I think.

Reply:
This is for future reference, currently that option should only be used by the `sources.**.enable` option implementation, hence keeping the previous description.

In the future this option may be useful for custom and/or third-party sources, so we'll likely take upstream's description into account.

I could remove the commented out upstream description if you like?

Reply:
Since the option is `visible = false` anyway, I've included both in the description and removed the commented out one.